From 078ab65396f4562d12a5d3db46db4808eddd298a Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Mon, 14 Nov 2022 11:39:52 +0800
Subject: [PATCH 0059/1398] [Huawei] Revert "memcg: support memcg sync reclaim
 work as kswapd"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187973

-------------------------------

This reverts commit 1496d67c65b1679341a4da6d3c9709c937be3d59.

A better solution of memcg async reclaim will be applied, just revert
origin memcg kswapd patchset.

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 mm/memcontrol.c |  4 ----
 mm/vmscan.c     | 40 ++--------------------------------------
 2 files changed, 2 insertions(+), 42 deletions(-)

diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 5612ff2ffb68..cc7368380f69 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -2368,10 +2368,8 @@ static void high_work_func(struct work_struct *work)
 {
 	struct mem_cgroup *memcg;
 
-	current->flags |= PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD;
 	memcg = container_of(work, struct mem_cgroup, high_work);
 	reclaim_high(memcg, MEMCG_CHARGE_BATCH, GFP_KERNEL);
-	current->flags &= ~(PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD);
 }
 
 /*
@@ -2541,11 +2539,9 @@ void mem_cgroup_handle_over_high(void)
 	 * memory.high is currently batched, whereas memory.max and the page
 	 * allocator run every time an allocation is made.
 	 */
-	current->flags |= PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD;
 	nr_reclaimed = reclaim_high(memcg,
 				    in_retry ? SWAP_CLUSTER_MAX : nr_pages,
 				    GFP_KERNEL);
-	current->flags &= ~(PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD);
 
 	/*
 	 * memory.high is breached and reclaim is unable to keep up. Throttle
diff --git a/mm/vmscan.c b/mm/vmscan.c
index 62b5b66c11be..70bc02f1e918 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -62,8 +62,6 @@
 #include <linux/page_cache_limit.h>
 #include "internal.h"
 
-#define MEMCG_KSWAPD_SCATOR 10
-
 #define CREATE_TRACE_POINTS
 #include <trace/events/vmscan.h>
 
@@ -2832,24 +2830,6 @@ static inline bool should_continue_reclaim(struct pglist_data *pgdat,
 	return inactive_lru_pages > pages_for_compaction;
 }
 
-static bool is_memcg_kswapd_stopped(struct scan_control *sc)
-{
-	struct mem_cgroup *memcg = sc->target_mem_cgroup;
-	bool is_stop = false;
-	unsigned long stop_flag = 0;
-
-	if (!cgroup_reclaim(sc))
-		return false;
-	if (memcg->memory.max == PAGE_COUNTER_MAX)
-		stop_flag = memcg->memory.high / 6;
-	else
-		stop_flag = memcg->memory.high - memcg->memory.max *
-			    MEMCG_KSWAPD_SCATOR / 1000;
-	is_stop = page_counter_read(&memcg->memory) < stop_flag;
-
-	return (current_is_kswapd() && is_stop);
-}
-
 static void shrink_node_memcgs(pg_data_t *pgdat, struct scan_control *sc)
 {
 	struct mem_cgroup *target_memcg = sc->target_mem_cgroup;
@@ -2904,14 +2884,6 @@ static void shrink_node_memcgs(pg_data_t *pgdat, struct scan_control *sc)
 			   sc->nr_scanned - scanned,
 			   sc->nr_reclaimed - reclaimed);
 
-		/*
-		 * Memcg background reclaim would break iter once memcg kswapd
-		 * flag is satisfied.
-		 */
-		if (is_memcg_kswapd_stopped(sc)) {
-			mem_cgroup_iter_break(target_memcg, memcg);
-			break;
-		}
 	} while ((memcg = mem_cgroup_iter(target_memcg, memcg, NULL)));
 }
 
@@ -3280,9 +3252,6 @@ static unsigned long do_try_to_free_pages(struct zonelist *zonelist,
 		__count_zid_vm_events(ALLOCSTALL, sc->reclaim_idx, 1);
 
 	do {
-		if (is_memcg_kswapd_stopped(sc))
-			break;
-
 		vmpressure_prio(sc->gfp_mask, sc->target_mem_cgroup,
 				sc->priority);
 		sc->nr_scanned = 0;
@@ -3345,13 +3314,8 @@ static unsigned long do_try_to_free_pages(struct zonelist *zonelist,
 		goto retry;
 	}
 
-	/*
-	 * Untapped cgroup reserves?  Don't OOM, retry.
-	 * memcg usage is lower than memory.high / 2, memcg kswapd will lead to
-	 * stop memcg reclaim, but should not break low protection.
-	 */
-	if (sc->memcg_low_skipped &&
-	    !(current_is_kswapd() && cgroup_reclaim(sc))) {
+	/* Untapped cgroup reserves?  Don't OOM, retry. */
+	if (sc->memcg_low_skipped) {
 		sc->priority = initial_priority;
 		sc->force_deactivate = 0;
 		sc->memcg_low_reclaim = 1;
-- 
2.17.1

